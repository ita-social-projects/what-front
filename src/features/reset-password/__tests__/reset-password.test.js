import React from 'react';
import { useSelector } from 'react-redux';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

import { useActions } from '@/shared/index.js';
import { resetPassword } from '@models/index.js';
import { ResetPassword } from '../reset-password';

jest.mock('react-redux', () => ({
  useSelector: jest.fn().mockReturnValue({
    isLoading: false,
    error: '',
    loaded: true,
  }),
}));

jest.mock('@/shared/hooks', () => ({
  useActions: jest.fn().mockReturnValue([jest.fn()]),
}));

describe('Reset password', () => {
  let formValues;
  beforeEach(() => {
    formValues = {
      email: 'testmail@gmail.com',
      newPassword: 'Testpassw!ord123',
      confirmPassword: 'Testpassw!ord123',
    };
  });

  it('should render component', () => {
    const { container } = render(
      <Router>
        <ResetPassword />
      </Router>
    );
    const resetPasswordContainer = container.getElementsByClassName(
      'container'
    );

    expect(resetPasswordContainer).toMatchSnapshot();
  });

  it('should validate the form', async () => {
    const { getByText } = render(
      <Router>
        <ResetPassword />
      </Router>
    );

    const submitButton = getByText('Confirm');

    await waitFor(() => {
      fireEvent.click(submitButton);
    });

    const error = getByText(
      'Fill the form to create a new password for your account'
    );

    expect(error).toBeInTheDocument();
  });

  it('should handle error receiving', () => {
    useSelector.mockReturnValue({
      isLoading: false,
      error: 'Something went wrong',
      loaded: true,
    });

    const { getByText } = render(
      <Router>
        <ResetPassword />
      </Router>
    );

    const error = getByText('Something went wrong');
    expect(error).toBeInTheDocument();
  });

  it('should submit the form with correct schema', async () => {
    const { getByPlaceholderText, getByText } = render(
      <Router>
        <ResetPassword />
      </Router>
    );

    const email = getByPlaceholderText('Email address');
    const newPassword = getByPlaceholderText('New password');
    const confirmPassword = getByPlaceholderText('Confirm password');
    const submitButton = getByText('Confirm');

    const [setResetPassword] = useActions(resetPassword);

    fireEvent.change(email, {
      target: {
        value: formValues.email,
      },
    });

    fireEvent.change(newPassword, {
      target: {
        value: formValues.newPassword,
      },
    });

    fireEvent.change(confirmPassword, {
      target: {
        value: formValues.confirmPassword,
      },
    });

    await waitFor(() => {
      fireEvent.click(submitButton);
    });

    setResetPassword({
      email: email.value,
      newPassword: newPassword.value,
      confirmPassword: confirmPassword.value,
    });

    expect(setResetPassword).toHaveBeenCalledWith(formValues);
  });
});
