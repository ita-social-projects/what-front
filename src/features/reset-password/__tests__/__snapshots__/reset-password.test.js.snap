// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reset password should render component 1`] = `
HTMLCollection [
  <div
    class="container"
  >
    <div
      class="row justify-content-center"
    >
      <div
        class="col-xl-6 col-md-8 col-sm-10 col-12"
      >
        <div
          class="card, shadow"
        >
          <form
            action="#"
            class="p-3"
            novalidate=""
          >
            <h3
              class="text-center"
            >
              Reset password
            </h3>
            <p
              class="text-center mt-3"
            >
              Fill the form to create a new password for your account
            </p>
            <hr />
            <div
              class="form-group"
            >
              <label
                for="email"
              >
                Email address
              </label>
              <input
                class="form-control"
                id="email"
                name="email"
                placeholder="Email address"
                type="email"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="newPassword"
              >
                New password
              </label>
              <input
                class="form-control"
                id="newPassword"
                name="newPassword"
                placeholder="New password"
                type="password"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="newPassword"
              >
                Confirm password
              </label>
              <input
                class="form-control"
                id="confirmNewPassword"
                name="confirmNewPassword"
                placeholder="Confirm password"
                type="password"
                value=""
              />
            </div>
            <div
              class="d-flex justify-content-center"
            >
              <button
                class="btn undefined"
                disabled=""
                type="submit"
              >
                Confirm
              </button>
            </div>
            
          </form>
        </div>
      </div>
    </div>
  </div>,
]
`;
